plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -413,35 +397,27 @@ compileExpr e = withContextM 2 (sdToTxt $ "Compiling expr:" GHC.<+> GHC.ppr e) $
-                pure $ PIR.Constant () $ PLC.someValue bs
+        -- IsString has only one method, so it's enough to know that it's an IsString method to know we're looking at fromString
+        -- We can safely commit to this match as soon as we've seen fromString - we won't accept any applications of fromString that aren't creating literals
+        -- of our builtin types.
+        (strip -> GHC.Var (GHC.idDetails -> GHC.ClassOpId cls)) `GHC.App` GHC.Type ty `GHC.App` _ `GHC.App` content | GHC.getName cls == GHC.isStringClassName ->
+            case GHC.tyConAppTyCon_maybe ty of
+                Just tc -> case stringExprContent (strip content) of
+                    Just bs ->
+                        if | GHC.getName tc == bsTyName     -> pure $ PIR.Constant () $ PLC.someValue bs
+                           | GHC.getName tc == stringTyName -> pure $ PIR.Constant () $ PLC.someValue $ TE.decodeUtf8 bs
+                           | otherwise -> throwSd UnsupportedError $ "Use of fromString on type other than builtin strings or bytestrings:" GHC.<+> GHC.ppr ty
+                    Nothing -> throwSd UnsupportedError $ "Use of fromString with inscrutable content:" GHC.<+> GHC.ppr content
+                Nothing -> throwSd UnsupportedError $ "Use of fromString on type other than builtin strings or bytestrings:" GHC.<+> GHC.ppr ty
         -- 'stringToBuiltinByteString' invocation, will be wrapped in a 'noinline'
-        (strip -> GHC.Var n) `GHC.App` (strip -> stringExprContent -> Just bs) | GHC.getName n == sbbsName -> do
+        (strip -> GHC.Var n) `GHC.App` (strip -> stringExprContent -> Just bs) | GHC.getName n == sbbsName ->
                 pure $ PIR.Constant () $ PLC.someValue bs
-        -- 'fromString' invocation at the builtin String type
-        (strip -> GHC.Var (GHC.idDetails -> GHC.ClassOpId cls)) `GHC.App` GHC.Type (GHC.tyConAppTyCon_maybe -> Just tc) `GHC.App` _ `GHC.App` (strip -> stringExprContent -> Just bs)
-            | GHC.getName cls == GHC.isStringClassName, GHC.getName tc == stringTyName -> do
-                let text = TE.decodeUtf8 bs
-                pure $ PIR.Constant () $ PLC.someValue text
         -- 'stringToBuiltinString' invocation, will be wrapped in a 'noinline'
-        (strip -> GHC.Var n) `GHC.App` (strip -> stringExprContent -> Just bs) | GHC.getName n == sbsName -> do
-                let text = TE.decodeUtf8 bs
-                pure $ PIR.Constant () $ PLC.someValue text
-        (strip -> GHC.Var (GHC.idDetails -> GHC.ClassOpId cls)) `GHC.App` GHC.Type ty `GHC.App` _ `GHC.App` _
-            | GHC.getName cls == GHC.isStringClassName ->
-                throwSd UnsupportedError $ "Use of fromString on type other than builtin strings or bytestrings:" GHC.<+> GHC.ppr ty
+        (strip -> GHC.Var n) `GHC.App` (strip -> stringExprContent -> Just bs) | GHC.getName n == sbsName ->
+                pure $ PIR.Constant () $ PLC.someValue $ TE.decodeUtf8 bs
```

Comment:
Do we want to use `decodeUtf8'` here and L408?

Reply:
Yes! I actually have no idea what happens if the plugin throws an exception, let's not find out :sweat_smile: